File: apple_watch_app/PathOfNurWatch/Refresh/PrayerPhaseEvaluator.swift
import Foundation

struct PrayerPhaseEvaluation {
    var state: PrayerPhaseState
    var didTransition: Bool
}

struct PrayerPhaseEvaluator {
    func evaluate(prayers: [WatchPrayerStatus], now: Date, previous: PrayerPhaseState?) -> PrayerPhaseEvaluation {
        let sorted = prayers.sorted { $0.scheduledTime < $1.scheduledTime }
        let next = sorted.first { $0.scheduledTime > now }
        let current = sorted.last { $0.scheduledTime <= now }
        let state = PrayerPhaseState(
            currentPrayerId: current?.prayerId,
            nextPrayerId: next?.prayerId,
            nextPrayerTime: next?.scheduledTime,
            lastEvaluatedAt: now
        )
        let didTransition = previous?.nextPrayerId != state.nextPrayerId || previous?.currentPrayerId != state.currentPrayerId
        return PrayerPhaseEvaluation(state: state, didTransition: didTransition)
    }
}

File: wear_os_app/app/src/main/java/com/pathofnur/watch/refresh/PrayerPhaseEvaluator.kt
package com.pathofnur.watch.refresh

import com.pathofnur.watch.domain.WatchPrayerStatus
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

data class PrayerPhaseEvaluation(
    val state: PrayerPhaseState,
    val didTransition: Boolean
)

class PrayerPhaseEvaluator {
    private val formatter = DateTimeFormatter.ofPattern("h:mm a").withZone(ZoneId.systemDefault())

    fun evaluate(prayers: List<WatchPrayerStatus>, nextPrayerId: String?, nextPrayerTime: Instant?, previous: PrayerPhaseState?): PrayerPhaseEvaluation {
        val next = prayers.firstOrNull { it.prayerId == nextPrayerId } ?: prayers.firstOrNull()
        val state = PrayerPhaseState(
            currentPrayerId = previous?.nextPrayerId,
            nextPrayerId = next?.prayerId,
            nextPrayerTimeLabel = nextPrayerTime?.let { formatter.format(it) } ?: next?.scheduledTimeLabel,
            lastEvaluatedAt = System.currentTimeMillis()
        )
        val didTransition = previous?.nextPrayerId != state.nextPrayerId
        return PrayerPhaseEvaluation(state, didTransition)
    }
}

